Fix matching of "components" dir when there's a symlink in the path to .platformio

mhkrebs opened the pull request:
`compile_source_files()` uses a relative path for the object if the source file is within the "components" tree.  It checks the path string with `.startswith()`, though, and that can fail if the paths point to the same file but look different due to symlinks.

In my case, it saw two different source paths in "target-__idf_efuse-5f3f9f6e3513e0903efa.json":
1. /space/home/user/.platformio/packages/framework-espidf/components/efuse/esp32c6/esp_efuse_fields.c
2. /space/home/user/.platformio/packages/framework-espidf/components/efuse/src/esp_efuse_fields.c

But my home directory uses a symlink, so `components_dir` was:
- /home/user/.platformio/packages/framework-espidf/components

This resulted in attempting to build the same object file for both:

```
*** Multiple ways to build the same target were specified for: /space/home/user/git/project/esphome/.esphome/build/device/.pioenvs/device/esp_efuse_fields.c.o  (from ['/space/home/user/.platformio/packages/framework-espidf/components/efuse/esp32c6/esp_efuse_fields.c'] and from ['/space/home/user/.platformio/packages/framework-espidf/components/efuse/src/esp_efuse_fields.c'])
File "/home/user/.platformio/platforms/espressif32/builder/frameworks/espidf.py", line 773, in compile_source_files
```

This fix simply resolves any symlinks in `components_dir` before using it for the comparison.

Fixes #515

mhkrebs committed:
Fix matching of "components" dir when there's a symlink in the path to .platformio

compile_source_files() uses a relative path for the object if the source file is within the "components" tree.  It checks the path string with .startswith(), though, and that can fail if the paths point to the same file but look different due to symlinks.

In my case, it saw two different source paths in target-__idf_efuse-5f3f9f6e3513e0903efa.json:
1. /space/home/user/.platformio/packages/framework-espidf/components/efuse/esp32c6/esp_efuse_fields.c
2. /space/home/user/.platformio/packages/framework-espidf/components/efuse/src/esp_efuse_fields.c

But my home directory uses a symlink, so components_dir was:
- /home/user/.platformio/packages/framework-espidf/components

This resulted in attempting to build the same object file for both:

*** Multiple ways to build the same target were specified for: /space/home/user/git/project/esphome/.esphome/build/device/.pioenvs/device/esp_efuse_fields.c.o  (from ['/space/home/user/.platformio/packages/framework-espidf/components/efuse/esp32c6/esp_efuse_fields.c'] and from ['/space/home/user/.platformio/packages/framework-espidf/components/efuse/src/esp_efuse_fields.c'])
File "/home/user/.platformio/platforms/espressif32/builder/frameworks/espidf.py", line 773, in compile_source_files

This fix simply resolves any symlinks in components_dir before using it for the comparison.

valeros commented:
Hi @mhkrebs, thanks for the PR. Could you please provide instructions how to reproduce the issue?

mhkrebs commented:
In my case (and in [this comment](https://github.com/platformio/platform-espressif32/issues/515#issuecomment-802927017)), the issue is that `/home` is a symlink.  (I'm on Linux).  I've seen others have the problem when `~/.platformio` is a symlink.

To reproduce maybe try:

```sh
cd
mv .platformio temp
ln -s temp .platformio
cd esphome
esphome compile device.yaml
```

(BTW, I don't feel sentimental about this PR... I can believe there's a better fix for this.)

mhkrebs committed:
Update my fix to work on Windows

My original fix resolved links in `components_dir`, but apparently the source paths might not be resolved themselves (at least on Windows).  This fix simply resolves links in both before doing the comparison.

mhkrebs commented:
Alright, I updated the fix.  I was worried about that possibility, that the source paths might not be resolved themselves.

I don't have Windows, but this still fixes my issue on Linux.  I therefore don't know how to test for Windows, but it looks like the tests will automatically run for this PR?  Although, like I implied, I'm fine if somebody wants to make a better fix for this... I just want to be able to compile ESPHome. :)

VulpesFurvus commented:
Hi, I'm running the device update in home assistant and having the same error with the latest update of ESPHome v2025.6.2:
```
INFO ESPHome 2025.6.2
INFO Reading configuration /config/esphome/m5stack-atom-lite-b85da4.yaml...
INFO Generating C++ source...
INFO Updating https://github.com/espressif/esp-protocols.git@mdns-v1.8.2
INFO Compiling app...
Processing m5stack-atom-lite-b85da4 (board: m5stack-atom; framework: espidf; platform: https://github.com/pioarduino/platform-espressif32/releases/download/53.03.13/platform-espressif32.zip)
--------------------------------------------------------------------------------
HARDWARE: ESP32 240MHz, 320KB RAM, 4MB Flash
 - framework-espidf @ 3.50302.0 (5.3.2) 
 - tool-cmake @ 3.30.2 
 - tool-esptoolpy @ 4.8.6 
 - tool-mklittlefs @ 3.2.0 
 - tool-ninja @ 1.10.2 
 - tool-riscv32-esp-elf-gdb @ 14.2.0+20240403 
 - tool-xtensa-esp-elf-gdb @ 14.2.0+20240403 
 - toolchain-esp32ulp @ 2.35.0-20220830 
 - toolchain-xtensa-esp-elf @ 13.2.0+20240530
Reading CMake configuration...
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_gpio/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_mm/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_hw_support/dma/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_hw_support/ldo/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/freertos/linker_common.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_gptimer/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_uart/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_pcnt/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_spi/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_mcpwm/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_ana_cmpr/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_dac/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_rmt/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_sdm/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_i2c/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_ledc/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_driver_parlio/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/driver/twai/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/vfs/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_netif/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_coex/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/bt/linker_common.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/bt/linker_rw_bt_controller.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_adc/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_eth/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/esp_psram/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/ieee802154/linker.lf` fragment is not found!
Warning! The `/data/cache/platformio/packages/framework-espidf/components/openthread/linker.lf` fragment is not found!

*** Multiple ways to build the same target were specified for: /data/build/m5stack-atom-lite-b85da4/.pioenvs/m5stack-atom-lite-b85da4/esp_efuse_fields.c.o  (from ['/data/cache/platformio/packages/framework-espidf/components/efuse/esp32/esp_efuse_fields.c'] and from ['/data/cache/platformio/packages/framework-espidf/components/efuse/src/esp_efuse_fields.c'])
File "/data/cache/platformio/platforms/espressif32/builder/frameworks/espidf.py", line 1044, in compile_source_files
========================== [FAILED] Took 6.73 seconds ==========================
```

HeadBoffin commented:
This issue just came up for me - any chance it will be merged in or do we have to manually update?

valeros commented:
Thanks, merged.
